### PR TITLE
gradient clipping in optimizers

### PR DESCRIPTION
**Pull request by hnyu**
For gin files that haven't been updated, I didn't change the gradient clipping options.

### CODE REVIEW

**Review by emailweixu — DISMISSED** [review]

**Review by Haichao-Zhang — DISMISSED** [review]

File: alf/optimizers/optimizers.py [hunk 1]
```diff
@@ -16,24 +16,75 @@
...
+                to clip gradient. If False, use `tensor_utils.clip_by_norms` for
+                each grad.
+            name (str): the name displayed when summarizing the gradient norm. If
+                None, then a global name in the format of "class_name_i" will be
+                created, where "i" is the global optimizer id.
+        """
         super(NewCls, self).__init__([{'params': []}], **kwargs)
+        self._gradient_clipping = gradient_clipping
+        self._clip_by_global_norm = clip_by_global_norm
+        self._name = name
```
Comment (Haichao-Zhang) [inline alf/optimizers/optimizers.py:55]: Is this ```name`` necessary?  Can use ```NewCls.\__name\__```?

Comment (hnyu) [reply alf/optimizers/optimizers.py:55]: If there are several Adam optimizers then their global norm summary will have conficts. The user then has the chance to specify their own name, if he doesn't want the default numbered name

Comment (Haichao-Zhang) [reply alf/optimizers/optimizers.py:55]: You can still keep the counter.

Comment (hnyu) [reply alf/optimizers/optimizers.py:55]: > You can still keep the counter.

Even the numbered names will be confusing like you don't know Adam_0 corresponds to your higher-level algorithm or your lower-level one. The order of creating optimizers can be uncertain or hard to find out.

**Comment by hnyu** [comment]
Resolved conflicts

File: alf/optimizers/optimizers.py (hunk 1 above)
Comment (Haichao-Zhang) [reply alf/optimizers/optimizers.py:55]: You are right. It's better to keep it if there are a lot of optimizers of the same type.

**Review by Haichao-Zhang — APPROVED** [review]

**Pull request merged**